extensions/amp-ad-exit/0.1/amp-ad-exit.js

```diff
@@ -187,7 +250,66 @@ export class AmpAdExit extends AMP.BaseElement {
       this.transport_.beacon = config.transport[TransportMode.BEACON] !== false;
       this.transport_.image = config.transport[TransportMode.IMAGE] !== false;
     } catch (e) {
-      this.user().error(TAG, 'Invalid JSON config', e);
+      user().error(TAG, 'Invalid JSON config', e);
+      throw e;
+    }
+
+    const ampAdResourceId = this.getAmpAdResourceId();
+
+    this.unlisten_ = listen(this.getAmpDoc().win, 'message', event => {
+      const responseMessage = deserializeMessage(getData(event));
+
+      this.assertValidResponseMessage(responseMessage, ampAdResourceId,
+          event.origin);
+
+      this.vendorResponses_[responseMessage['vendor']] =
+          responseMessage['message'];
+    });
+  }
+
+  /**
+   *
+   * @param responseMessage The response object to validate.
+   * @param expectedCreativeId The resource ID of the enclosing AMP ad (which
+   *     responseMessage['creativeId'] should match.
+   * @param expectedVendor The 3p analytics vendor, which
+   *     responseMesssage['vendor'] should match.
+   */
+  assertValidResponseMessage(responseMessage, expectedCreativeId,
+                             expectedVendor) {
+    if (!responseMessage || !responseMessage['type'] ||
+      responseMessage['type'] != MessageType.IFRAME_TRANSPORT_RESPONSE ||
+      !responseMessage['creativeId'] ||
+      responseMessage['creativeId'] != expectedCreativeId) {
+      return;
+    }
+    dev().assert(responseMessage && responseMessage['message'],
+        'Received empty response from 3p analytics frame');
+    dev().assert(responseMessage['type'] &&
+        responseMessage['type'] == MessageType.IFRAME_TRANSPORT_RESPONSE,
+        'Received response message of invalid type from 3p analytics frame');
+    dev().assert(responseMessage['creativeId'] &&
+        responseMessage['creativeId'] == expectedCreativeId,
+        'Received malformed message from 3p analytics frame: ' +
+        'creativeId missing');
+    dev().assert(responseMessage['vendor'],
+        'Received malformed message from 3p analytics frame: ' +
+        'vendor missing');
+    assertOriginMatchesVendor(expectedVendor, responseMessage['vendor']);
+  }
+
+  /**
+   * Gets the resource ID of the amp-ad element containing this AmpAdExit
+   * instance.
+   * @return {string}
+   */
+  getAmpAdResourceId() {
```

Comment:
This is essentially the same as in amp-analytics, only the comments and error message text differ.
Any suggestions for a common place to put this? src/dom?